File: python/design_pattern/facade/main_facade.py
# /// script
# requires-python = ">=3.10"
# dependencies = []
# ///
"""
Facade Pattern

The Facade pattern provides a unified interface to a set of interfaces in a
subsystem. It defines a higher-level interface that makes the subsystem easier
to use.

Key Components:
- Facade: Provides simple interface to complex subsystem
- Subsystem classes: Implement subsystem functionality
- Client: Uses Facade instead of subsystem directly
"""


# Subsystem classes for a home theater system
class Amplifier:
    """Audio amplifier subsystem."""

    def __init__(self, description: str):
        self.description = description
        self._volume = 0
        self._source = ""

    def on(self) -> str:
        return f"{self.description} is ON"

    def off(self) -> str:
        return f"{self.description} is OFF"

    def set_source(self, source: str) -> str:
        self._source = source
        return f"{self.description} source set to {source}"

    def set_volume(self, level: int) -> str:
        self._volume = level
        return f"{self.description} volume set to {level}"

    def set_surround_sound(self) -> str:
        return f"{self.description} surround sound ON"


class DvdPlayer:
    """DVD player subsystem."""

    def __init__(self, description: str):
        self.description = description
        self._movie = ""

    def on(self) -> str:
        return f"{self.description} is ON"

    def off(self) -> str:
        return f"{self.description} is OFF"

    def play(self, movie: str) -> str:
        self._movie = movie
        return f"{self.description} playing '{movie}'"

    def pause(self) -> str:
        return f"{self.description} paused '{self._movie}'"

    def stop(self) -> str:
        return f"{self.description} stopped '{self._movie}'"

    def eject(self) -> str:
        return f"{self.description} DVD ejected"


class StreamingPlayer:
    """Streaming player subsystem."""

    def __init__(self, description: str):
        self.description = description
        self._show = ""

    def on(self) -> str:
        return f"{self.description} is ON"

    def off(self) -> str:
        return f"{self.description} is OFF"

    def play(self, show: str) -> str:
        self._show = show
        return f"{self.description} streaming '{show}'"

    def stop(self) -> str:
        return f"{self.description} stopped streaming"


class Projector:
    """Projector subsystem."""

    def __init__(self, description: str):
        self.description = description

    def on(self) -> str:
        return f"{self.description} is ON"

    def off(self) -> str:
        return f"{self.description} is OFF"

    def wide_screen_mode(self) -> str:
        return f"{self.description} in widescreen mode (16:9)"

    def tv_mode(self) -> str:
        return f"{self.description} in TV mode (4:3)"


class Screen:
    """Projection screen subsystem."""

    def __init__(self, description: str):
        self.description = description

    def down(self) -> str:
        return f"{self.description} going down"

    def up(self) -> str:
        return f"{self.description} going up"


class TheaterLights:
    """Theater lighting subsystem."""

    def __init__(self, description: str):
        self.description = description
        self._level = 100

    def on(self) -> str:
        return f"{self.description} are ON"

    def off(self) -> str:
        return f"{self.description} are OFF"

    def dim(self, level: int) -> str:
        self._level = level
        return f"{self.description} dimmed to {level}%"


class PopcornPopper:
    """Popcorn maker subsystem."""

    def __init__(self, description: str):
        self.description = description

    def on(self) -> str:
        return f"{self.description} is ON"

    def off(self) -> str:
        return f"{self.description} is OFF"

    def pop(self) -> str:
        return f"{self.description} popping popcorn!"


# Facade
class HomeTheaterFacade:
    """Facade providing simple interface to home theater subsystem."""

    def __init__(
        self,
        amp: Amplifier,
        dvd: DvdPlayer,
        streaming: StreamingPlayer,
        projector: Projector,
        screen: Screen,
        lights: TheaterLights,
        popper: PopcornPopper,
    ):
        self.amp = amp
        self.dvd = dvd
        self.streaming = streaming
        self.projector = projector
        self.screen = screen
        self.lights = lights
        self.popper = popper

    def watch_movie(self, movie: str) -> list[str]:
        """Simple operation to watch a DVD movie."""
        actions = [
            "Get ready to watch a movie...",
            self.popper.on(),
            self.popper.pop(),
            self.lights.dim(10),
            self.screen.down(),
            self.projector.on(),
            self.projector.wide_screen_mode(),
            self.amp.on(),
            self.amp.set_source("DVD"),
            self.amp.set_surround_sound(),
            self.amp.set_volume(5),
            self.dvd.on(),
            self.dvd.play(movie),
        ]
        return actions

    def end_movie(self) -> list[str]:
        """Simple operation to end movie watching."""
        actions = [
            "Shutting movie theater down...",
            self.popper.off(),
            self.lights.on(),
            self.screen.up(),
            self.projector.off(),
            self.amp.off(),
            self.dvd.stop(),
            self.dvd.eject(),
            self.dvd.off(),
        ]
        return actions

    def stream_show(self, show: str) -> list[str]:
        """Simple operation to stream a show."""
        actions = [
            "Get ready to stream a show...",
            self.lights.dim(20),
            self.projector.on(),
            self.projector.tv_mode(),
            self.amp.on(),
            self.amp.set_source("Streaming"),
            self.amp.set_volume(4),
            self.streaming.on(),
            self.streaming.play(show),
        ]
        return actions

    def end_streaming(self) -> list[str]:
        """Simple operation to end streaming."""
        actions = [
            "Shutting down streaming...",
            self.streaming.stop(),
            self.streaming.off(),
            self.amp.off(),
            self.projector.off(),
            self.lights.on(),
        ]
        return actions

    def listen_to_music(self, source: str = "Bluetooth") -> list[str]:
        """Simple operation to listen to music."""
        actions = [
            "Setting up for music...",
            self.lights.dim(50),
            self.amp.on(),
            self.amp.set_source(source),
            self.amp.set_volume(6),
        ]
        return actions

    def end_music(self) -> list[str]:
        """Simple operation to stop music."""
        actions = [
            "Shutting down music...",
            self.amp.off(),
            self.lights.on(),
        ]
        return actions


# Additional facade example: Computer startup
class CPU:
    def freeze(self) -> str:
        return "CPU: Freezing processor"

    def jump(self, address: str) -> str:
        return f"CPU: Jumping to {address}"

    def execute(self) -> str:
        return "CPU: Executing instructions"


class Memory:
    def load(self, address: str, data: str) -> str:
        return f"Memory: Loading '{data}' at {address}"


class HardDrive:
    def read(self, sector: int, size: int) -> str:
        return f"HardDrive: Reading {size} bytes from sector {sector}"


class ComputerFacade:
    """Facade for computer startup."""

    def __init__(self):
        self.cpu = CPU()
        self.memory = Memory()
        self.hard_drive = HardDrive()

    def start(self) -> list[str]:
        """Simple start operation."""
        return [
            "Starting computer...",
            self.cpu.freeze(),
            self.memory.load("0x00", "BIOS"),
            self.cpu.jump("0x00"),
            self.hard_drive.read(0, 1024),
            self.memory.load("0x0A", "Boot loader"),
            self.cpu.jump("0x0A"),
            self.cpu.execute(),
            "Computer started successfully!",
        ]


def main() -> None:
    print("=" * 60)
    print("Facade Pattern - Home Theater Demo")
    print("=" * 60)

    # Create subsystem components
    amp = Amplifier("Top-O-Line Amplifier")
    dvd = DvdPlayer("Top-O-Line DVD Player")
    streaming = StreamingPlayer("Streaming Player")
    projector = Projector("Top-O-Line Projector")
    screen = Screen("Theater Screen")
    lights = TheaterLights("Theater Ceiling Lights")
    popper = PopcornPopper("Popcorn Popper")

    # Create facade
    home_theater = HomeTheaterFacade(
        amp, dvd, streaming, projector, screen, lights, popper
    )

    # Demo 1: Watch a movie (simple operation)
    print("\n--- Watch Movie ---")
    for action in home_theater.watch_movie("Raiders of the Lost Ark"):
        print(f"  {action}")

    print("\n--- End Movie ---")
    for action in home_theater.end_movie():
        print(f"  {action}")

    # Demo 2: Stream a show
    print("\n--- Stream Show ---")
    for action in home_theater.stream_show("Stranger Things"):
        print(f"  {action}")

    print("\n--- End Streaming ---")
    for action in home_theater.end_streaming():
        print(f"  {action}")

    # Demo 3: Listen to music
    print("\n--- Listen to Music ---")
    for action in home_theater.listen_to_music("Spotify"):
        print(f"  {action}")

    print("\n--- End Music ---")
    for action in home_theater.end_music():
        print(f"  {action}")

    # Demo 4: Computer facade
    print("\n--- Computer Startup Facade ---")
    computer = ComputerFacade()
    for action in computer.start():
        print(f"  {action}")

    # Demo 5: Show the complexity hidden by facade
    print("\n--- Without Facade (Direct Subsystem Access) ---")
    print("To watch a movie, client would need to:")
    print("  1. popper.on()")
    print("  2. popper.pop()")
    print("  3. lights.dim(10)")
    print("  4. screen.down()")
    print("  5. projector.on()")
    print("  6. projector.wide_screen_mode()")
    print("  7. amp.on()")
    print("  8. amp.set_source('DVD')")
    print("  9. amp.set_surround_sound()")
    print("  10. amp.set_volume(5)")
    print("  11. dvd.on()")
    print("  12. dvd.play(movie)")
    print("\nWith Facade: home_theater.watch_movie(movie)")

    print("\n" + "=" * 60)
    print("Benefits of Facade Pattern:")
    print("=" * 60)
    print("1. Simplifies client interaction with complex subsystems")
    print("2. Decouples client from subsystem components")
    print("3. Doesn't prevent direct subsystem access if needed")
    print("4. Can layer facades for large systems")
    print("5. Follows Principle of Least Knowledge")


if __name__ == "__main__":
    main()

File: java/design_pattern/facade/MainFacade.java
/**
 * Comprehensive demonstration of the Facade Pattern in Java
 *
 * The Facade pattern provides a unified interface to a set of interfaces
 * in a subsystem. It defines a higher-level interface that makes the
 * subsystem easier to use.
 */

// Complex subsystem classes for Home Theater
class Amplifier {
    private String description;
    private int volume;

    public Amplifier(String description) {
        this.description = description;
    }

    public void on() {
        System.out.println("  [Amplifier] " + description + " is ON");
    }

    public void off() {
        System.out.println("  [Amplifier] " + description + " is OFF");
    }

    public void setVolume(int level) {
        this.volume = level;
        System.out.println("  [Amplifier] Setting volume to " + level);
    }

    public void setSurroundSound() {
        System.out.println("  [Amplifier] Surround sound enabled");
    }

    public void setStereoSound() {
        System.out.println("  [Amplifier] Stereo mode enabled");
    }
}

class DVDPlayer {
    private String description;
    private String currentMovie;

    public DVDPlayer(String description) {
        this.description = description;
    }

    public void on() {
        System.out.println("  [DVDPlayer] " + description + " is ON");
    }

    public void off() {
        System.out.println("  [DVDPlayer] " + description + " is OFF");
    }

    public void play(String movie) {
        this.currentMovie = movie;
        System.out.println("  [DVDPlayer] Playing \"" + movie + "\"");
    }

    public void stop() {
        System.out.println("  [DVDPlayer] Stopped \"" + currentMovie + "\"");
    }

    public void pause() {
        System.out.println("  [DVDPlayer] Paused \"" + currentMovie + "\"");
    }

    public void eject() {
        System.out.println("  [DVDPlayer] Ejecting disc");
    }
}

class Projector {
    private String description;

    public Projector(String description) {
        this.description = description;
    }

    public void on() {
        System.out.println("  [Projector] " + description + " is ON");
    }

    public void off() {
        System.out.println("  [Projector] " + description + " is OFF");
    }

    public void wideScreenMode() {
        System.out.println("  [Projector] Widescreen mode (16:9)");
    }

    public void tvMode() {
        System.out.println("  [Projector] TV mode (4:3)");
    }
}

class TheaterLights {
    private String description;

    public TheaterLights(String description) {
        this.description = description;
    }

    public void on() {
        System.out.println("  [Lights] " + description + " ON");
    }

    public void off() {
        System.out.println("  [Lights] " + description + " OFF");
    }

    public void dim(int level) {
        System.out.println("  [Lights] Dimming to " + level + "%");
    }
}

class Screen {
    private String description;

    public Screen(String description) {
        this.description = description;
    }

    public void up() {
        System.out.println("  [Screen] " + description + " going up");
    }

    public void down() {
        System.out.println("  [Screen] " + description + " going down");
    }
}

class PopcornMachine {
    private String description;

    public PopcornMachine(String description) {
        this.description = description;
    }

    public void on() {
        System.out.println("  [Popcorn] " + description + " is ON");
    }

    public void off() {
        System.out.println("  [Popcorn] " + description + " is OFF");
    }

    public void pop() {
        System.out.println("  [Popcorn] Popping popcorn!");
    }
}

// FACADE - simplifies the complex subsystem
class HomeTheaterFacade {
    private Amplifier amp;
    private DVDPlayer dvd;
    private Projector projector;
    private TheaterLights lights;
    private Screen screen;
    private PopcornMachine popcorn;

    public HomeTheaterFacade(Amplifier amp, DVDPlayer dvd, Projector projector,
                             TheaterLights lights, Screen screen, PopcornMachine popcorn) {
        this.amp = amp;
        this.dvd = dvd;
        this.projector = projector;
        this.lights = lights;
        this.screen = screen;
        this.popcorn = popcorn;
    }

    public void watchMovie(String movie) {
        System.out.println("Get ready to watch a movie...\n");
        popcorn.on();
        popcorn.pop();
        lights.dim(10);
        screen.down();
        projector.on();
        projector.wideScreenMode();
        amp.on();
        amp.setSurroundSound();
        amp.setVolume(5);
        dvd.on();
        dvd.play(movie);
    }

    public void endMovie() {
        System.out.println("\nShutting movie theater down...\n");
        popcorn.off();
        lights.on();
        screen.up();
        projector.off();
        amp.off();
        dvd.stop();
        dvd.eject();
        dvd.off();
    }

    public void listenToMusic(String track) {
        System.out.println("Setting up for music...\n");
        lights.dim(50);
        amp.on();
        amp.setStereoSound();
        amp.setVolume(8);
        System.out.println("  Playing track: " + track);
    }
}

// Another example: Computer startup

class CPU {
    public void freeze() {
        System.out.println("  [CPU] Freezing processor");
    }

    public void jump(long position) {
        System.out.println("  [CPU] Jumping to position " + position);
    }

    public void execute() {
        System.out.println("  [CPU] Executing instructions");
    }
}

class Memory {
    public void load(long position, byte[] data) {
        System.out.println("  [Memory] Loading data at position " + position);
    }
}

class HardDrive {
    public byte[] read(long lba, int size) {
        System.out.println("  [HardDrive] Reading " + size + " bytes from sector " + lba);
        return new byte[size];
    }
}

// Computer Facade
class ComputerFacade {
    private static final long BOOT_ADDRESS = 0x0000;
    private static final long BOOT_SECTOR = 0;
    private static final int SECTOR_SIZE = 512;

    private CPU cpu;
    private Memory memory;
    private HardDrive hardDrive;

    public ComputerFacade() {
        this.cpu = new CPU();
        this.memory = new Memory();
        this.hardDrive = new HardDrive();
    }

    public void start() {
        System.out.println("Starting computer...\n");
        cpu.freeze();
        byte[] bootData = hardDrive.read(BOOT_SECTOR, SECTOR_SIZE);
        memory.load(BOOT_ADDRESS, bootData);
        cpu.jump(BOOT_ADDRESS);
        cpu.execute();
        System.out.println("\n  Computer started successfully!");
    }
}

// Third example: Order processing

class Inventory {
    public boolean checkStock(String productId) {
        System.out.println("  [Inventory] Checking stock for " + productId);
        return true; // Simplified
    }

    public void updateStock(String productId, int quantity) {
        System.out.println("  [Inventory] Updating stock: " + productId + " -" + quantity);
    }
}

class Payment {
    public boolean processPayment(String orderId, double amount) {
        System.out.println("  [Payment] Processing $" + String.format("%.2f", amount) + " for order " + orderId);
        return true; // Simplified
    }

    public void refund(String orderId) {
        System.out.println("  [Payment] Refunding order " + orderId);
    }
}

class Shipping {
    public String createShipment(String orderId, String address) {
        System.out.println("  [Shipping] Creating shipment for " + orderId + " to " + address);
        return "TRACK-" + orderId;
    }

    public void cancelShipment(String trackingId) {
        System.out.println("  [Shipping] Cancelling shipment " + trackingId);
    }
}

class Notification {
    public void sendOrderConfirmation(String email, String orderId) {
        System.out.println("  [Notification] Sending confirmation to " + email + " for " + orderId);
    }

    public void sendShippingUpdate(String email, String trackingId) {
        System.out.println("  [Notification] Sending shipping update to " + email + ": " + trackingId);
    }
}

// Order Facade
class OrderFacade {
    private Inventory inventory;
    private Payment payment;
    private Shipping shipping;
    private Notification notification;

    public OrderFacade() {
        this.inventory = new Inventory();
        this.payment = new Payment();
        this.shipping = new Shipping();
        this.notification = new Notification();
    }

    public boolean placeOrder(String orderId, String productId, int quantity,
                               double amount, String email, String address) {
        System.out.println("Processing order " + orderId + "...\n");

        // Check inventory
        if (!inventory.checkStock(productId)) {
            System.out.println("  Order failed: Out of stock");
            return false;
        }

        // Process payment
        if (!payment.processPayment(orderId, amount)) {
            System.out.println("  Order failed: Payment declined");
            return false;
        }

        // Update inventory
        inventory.updateStock(productId, quantity);

        // Create shipment
        String trackingId = shipping.createShipment(orderId, address);

        // Send notifications
        notification.sendOrderConfirmation(email, orderId);
        notification.sendShippingUpdate(email, trackingId);

        System.out.println("\n  Order " + orderId + " completed successfully!");
        return true;
    }
}

public class MainFacade {
    public static void main(String[] args) {
        System.out.println("=== Facade Pattern Demonstration ===\n");

        // Home Theater example
        System.out.println("--- 1. Home Theater Facade ---");

        Amplifier amp = new Amplifier("Top-O-Line Amplifier");
        DVDPlayer dvd = new DVDPlayer("Sony DVD Player");
        Projector projector = new Projector("Epson HD Projector");
        TheaterLights lights = new TheaterLights("Theater Ceiling Lights");
        Screen screen = new Screen("120-inch Screen");
        PopcornMachine popcorn = new PopcornMachine("Popcorn Maker");

        HomeTheaterFacade homeTheater = new HomeTheaterFacade(
            amp, dvd, projector, lights, screen, popcorn
        );

        homeTheater.watchMovie("Inception");
        homeTheater.endMovie();
        System.out.println();

        // Computer startup example
        System.out.println("--- 2. Computer Startup Facade ---");
        ComputerFacade computer = new ComputerFacade();
        computer.start();
        System.out.println();

        // Order processing example
        System.out.println("--- 3. E-Commerce Order Facade ---");
        OrderFacade orderSystem = new OrderFacade();
        orderSystem.placeOrder(
            "ORD-12345",
            "PROD-789",
            2,
            99.99,
            "customer@example.com",
            "123 Main St, City, State 12345"
        );

        System.out.println("\n=== Summary ===");
        System.out.println("Facade pattern benefits:");
        System.out.println("  - Simplifies complex subsystem interface");
        System.out.println("  - Reduces dependencies on subsystem internals");
        System.out.println("  - Wraps poorly-designed APIs with better ones");
        System.out.println("  - Provides context-specific interface to generic functionality");
        System.out.println("\nKey points:");
        System.out.println("  - Facade doesn't prevent direct subsystem access if needed");
        System.out.println("  - Can have multiple facades for different use cases");
        System.out.println("  - Subsystem classes are unaware of facade");
    }
}
